static/js/tracking.ts

```diff
@@ -0,0 +1,231 @@
+    const currentTime = Date.now();
+    const inactiveDuration = currentTime - lastActiveTime;
+    if (inactiveDuration <= 200) {
+        return;
+    }
+    const target = event.target as HTMLElement;
+    // console.log(target, event.type)
+    if (target.matches('button') || target.matches('a') || target.matches('input') || target.matches('select') || target.matches("div")) {
+        let elementIdOrName = target.id;
+
+        if (!elementIdOrName && target.hasAttribute("name")) {
+            elementIdOrName = target.getAttribute("name") || "";
+        }
+
+        if (ELEMENT_TO_TRACK.includes(elementIdOrName)) {
+            clickCounts = handleLocalStorage(CLICK_COUNTS);
+            
+            const page = window.location.pathname;
+
+            const value = (target as HTMLInputElement).value || "";
+
+            clickCounts.push({time: currentTime, id: elementIdOrName, page, extra: value});
+            
+            // console.log(target, clickCounts)
+            // console.log(`Event: ${event.type}, Element ID or Name: ${elementIdOrName}, Click Count: ${clickCounts}`);
+            handleUserActivity(clickCounts);
+        } 
+    }
+}
+
+// Function to handle user activity
+function handleUserActivity(clickCounts: any) {
+    lastActiveTime = handleLocalStorage(LAST_ACTIVE, Date.now());
+    clickCounts = handleLocalStorage(CLICK_COUNTS, clickCounts);
+    changesSent = false;
+}
+
+// Retrieve or set items in local storage.
+function handleLocalStorage(item: string, value: any = undefined) {
+    const retrievedItem = window.localStorage.getItem(item);
+    if (!retrievedItem || value !== undefined) {
+        value = value || [];
+        window.localStorage.setItem(item, JSON.stringify(value));
+    } else {
+        if (item === CLICK_COUNTS) {
+            value = JSON.parse(retrievedItem);
+        } else if (item === LAST_ACTIVE) {
+            value = parseInt(retrievedItem);
+        }
+    }
+    return value;
+}
+
+// Function to check user activity and send request if inactive for 5 minutes
+async function checkUserActivity() {
+    if (changesSent) {
+        // Perhaps add current page with no action by the user.
+        // clickCounts = handleLocalStorage(CLICK_COUNTS);
+        // const page = window.location.pathname;
+        // clickCounts.push({time: lastActiveTime, id: '', page});
+        // handleUserActivity(clickCounts);
+        return;
+    }
+    const currentTime = Date.now();
+    const inactiveDuration = currentTime - lastActiveTime;
+    if (inactiveDuration >= WAITING_TIME) {
+        sendRequestToServer();
+    }
+}
+
+// Function to send request to the server
+async function sendRequestToServer() {
+    try {
+        const data = handleLocalStorage(CLICK_COUNTS)
+        if (data.length) {
+            // console.log('Sending request to server...');
+            await postJson('/tracking', data);
+            handleUserActivity([]);
```

Comment:
Because this function is `async`, there is a race condition here. The following sequence of events can happen:

1. We read the current set of clicks from LocalStorage
2. We send them off to the server (this is async, so other code can run in the mean time)
3. User interacts with the page, adding a new event to the array
4. The response comes back, this function resumes
5. We clear the LocalStorage array

We've now dropped the additional user event on step 3.

To fix this, we can store the length of the array that we've sent off to the server, and when we go to clear the array do `array.splice(0, amountOfRecordsWeSent)`, so that if new ones got added in the mean time they don't get dropped.

Do we need `changesSent`? When we send changes, we reset the array to empty, so the fact whether the array is empty or non-empty can by itself be a signal for whether there are things to send or not.